core/src/main/java/org/elasticsearch/common/network/NetworkModule.java

```diff
@@ -19,21 +19,362 @@
+        this.settings = settings;
+        this.transportClient = transportClient;
+        registerTransportService(NETTY_TRANSPORT, TransportService.class);
+        registerTransport(LOCAL_TRANSPORT, LocalTransport.class);
+        registerTransport(NETTY_TRANSPORT, NettyTransport.class);
+
+        if (transportClient == false) {
+            registerHttpTransport(NETTY_TRANSPORT, NettyHttpServerTransport.class);
+
+            for (Class<? extends AbstractCatAction> catAction : builtinCatHandlers) {
+                catHandlers.registerExtension(catAction);
+            }
+            for (Class<? extends RestHandler> restAction : builtinRestHandlers) {
+                restHandlers.registerExtension(restAction);
+            }
+        }
+    }
+
+    /** Adds a transport service implementation that can be selected by setting {@link #TRANSPORT_SERVICE_TYPE_KEY}. */
+    public void registerTransportService(String name, Class<? extends TransportService> clazz) {
+        transportServiceTypes.registerExtension(name, clazz);
+    }
+
+    /** Adds a transport implementation that can be selected by setting {@link #TRANSPORT_TYPE_KEY}. */
+    public void registerTransport(String name, Class<? extends Transport> clazz) {
+        transportTypes.registerExtension(name, clazz);
+    }
+
+    /** Adds an http transport implementation that can be selected by setting {@link #HTTP_TYPE_KEY}. */
+    // TODO: we need another name than "http transport"....so confusing with transportClient...
+    public void registerHttpTransport(String name, Class<? extends HttpServerTransport> clazz) {
+        if (transportClient) {
+            throw new IllegalArgumentException("Cannot register http transport " + clazz.getName() + " for transport client");
+        }
+        httpTransportTypes.registerExtension(name, clazz);
+    }
+
+    /** Adds an additional rest action. */
+    // TODO: change this further to eliminate the middle man, ie RestController, and just register method and path here
+    public void registerRestHandler(Class<? extends RestHandler> clazz) {
+        if (transportClient) {
+            throw new IllegalArgumentException("Cannot register rest handler " + clazz.getName() + " for transport client");
+        }
+        if (AbstractCatAction.class.isAssignableFrom(clazz)) {
+            catHandlers.registerExtension(clazz.asSubclass(AbstractCatAction.class));
+        } else {
+            restHandlers.registerExtension(clazz);
+        }
     }
 
     @Override
     protected void configure() {
         bind(NetworkService.class).toInstance(networkService);
+        bind(NamedWriteableRegistry.class).asEagerSingleton();
+
+        transportServiceTypes.bindType(binder(), settings, TRANSPORT_SERVICE_TYPE_KEY, NETTY_TRANSPORT);
+        String defaultTransport = DiscoveryNode.localNode(settings) ? LOCAL_TRANSPORT : NETTY_TRANSPORT;
+        transportTypes.bindType(binder(), settings, TRANSPORT_TYPE_KEY, defaultTransport);
```

Comment:
> we can open a new issue to add debug logging to SelectedType

Agreed. I think it's a good one, but just to be cleared - it's not what I requested.

> but this is really arbitrary here

To me this is a case by case call of what to log on what level. To me changing a transport layer is something that's worth logging under info.

>  we have enough info going on and if we reset it we see it in the settings, no need to X different ways to get some information

It's a valid point that now it's visible in the settings and it's great. Logs are different and some people look at them (at least I did) to see things started up properly.

All of that said I don't under stand why we're discussing this to death. To *me* it would be nice to have this in the log. I'm sure other would as well. It's a low risk, one line change. Unless anyone is strongly against it, I don't understand what prevents us from just taking the feedback and a request for a minor trivial change. The amount of discussion - unless I miss a critical point - is not proportional to what I ask.